### PR TITLE
Make cross-server RPCs more reliable

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Better randomness by abusing FGuids, and larger buffer by increasing retention to 30 seconds.
Both of these changes worsen performance, but I don't think it will be too significant.

#### Release note
N/A ?

#### Tests
Ran tests a LOT of times.

#### Documentation
N/A ?

#### Primary reviewers
idk yet

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h [hunk 1]
```diff
@@ -13,7 +13,7 @@ class SpatialEventTracer;
 DECLARE_DELEGATE_RetVal_OneParam(bool, FProcessCrossServerRPC, const FCrossServerRPCParams&);
 DECLARE_DELEGATE_RetVal_OneParam(FCrossServerRPCParams, FTryRetrieveCrossServerRPCParams, const Worker_CommandRequestOp&);
 
-static double CrossServerRPCGuidTimeout = 5.f;
+static double CrossServerRPCGuidTimeout = 30.f;
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h:16]: Driveby - why is there a static double at in a header? I presume this is meant to be a constant?

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h:16]: hm, good point
can make it a constexpr. Would you like to move it as well? Could also see this being an advanced SpatialGDK setting.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h:16]: We can do if we think it's something users need to change but otherwise I think it should be a static const field on teh rpc handler (rather than global).

Also why 30 seconds?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp [hunk 2]
```diff
@@ -27,8 +27,8 @@ void CrossServerRPCSender::SendCommand(const FUnrealObjectRef InTargetObjectRef,
 
 	CommandRequest CommandRequest(SpatialConstants::SERVER_TO_SERVER_COMMAND_ENDPOINT_COMPONENT_ID,
 								  SpatialConstants::UNREAL_RPC_ENDPOINT_COMMAND_ID);
-
-	uint64 UniqueRPCId = FMath::RandHelper(INT_MAX);
+	FGuid Guid = FGuid::NewGuid();
+	uint64 UniqueRPCId = (((uint64)Guid.C) << 32) + (uint64)Guid.D;
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:31]: Worth considering using an actual high quality random number generator? PCG is pretty small?

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:31]: and really performant.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:31]: This entire implementation is kind of stop-gap, as it will never actually be perfect. There is also a crypto lib currently in the engine that seems to do a pretty good random, but I didn't want to introduce the dependency just because of this.
This GUID call seems to resolve to some OS-internal function, but I can research a bit if it's high-quality enough for us.
(And tbf, the usage here - to generate ids for rpcs seems better suited to guids anyway, just concerned with throwing away the higher bits).

File: SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h:16]: I have found 5 to be within the realms of unreliability and random stalls. This is just a finger-in-the-air number, but I thought 30.0 would be better. Can vote/discuss.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h:16]: (as in, I got a 5s stall while testing this that caused issues)

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/CrossServerRPCHandler.h]: Moved into cpp and made constexpr

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp (hunk 2 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:31]: I checked a bit - on windows it's something really opaque. As a fallback on an unknown platform it's mostly system time + cycles in the lower bits, so that is not great. But the actual implementation on unix is using urandom (or at least tries to).

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp [hunk 3]
```diff
@@ -10,6 +10,8 @@ DEFINE_LOG_CATEGORY(LogCrossServerRPCHandler);
 
 namespace SpatialGDK
 {
+static constexpr double CrossServerRPCGuidTimeout = 30.f;
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp:13]: Better still would be to put it in the class if we don't need it elsewhere. But this is still better.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp:13]: hmm, not sure if I understand. Like `CrossServerRPCHandler::CrossServerRPCGuidTimeout = 30.f`? And make it private?

**Comment by mironec** [comment]
Small update, made use of all the bits of a guid and moved the constexpr inside the class header.

**Comment by mironec** [comment]
@m-samiec do we want to actually merge this? Or do you want to go ahead with removing the ids altogether? I think that would take some more time and more testing/theorycrafting.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp [hunk 4]
```diff
@@ -28,7 +28,9 @@ void CrossServerRPCSender::SendCommand(const FUnrealObjectRef InTargetObjectRef,
 	CommandRequest CommandRequest(SpatialConstants::SERVER_TO_SERVER_COMMAND_ENDPOINT_COMPONENT_ID,
 								  SpatialConstants::UNREAL_RPC_ENDPOINT_COMMAND_ID);
 
-	uint64 UniqueRPCId = FMath::RandHelper(INT_MAX);
+	// We just want a semi-globally unique id for the RPC to avoid false discards.
+	FGuid Guid = FGuid::NewGuid();
+	uint64 UniqueRPCId = (((uint64)Guid.A ^ Guid.B) << 32) + (uint64)(Guid.C ^ Guid.D);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:33]: Could use `CityHash64` here instead.

**Comment by m-samiec** [comment]
> @m-samiec do we want to actually merge this? Or do you want to go ahead with removing the ids altogether? I think that would take some more time and more testing/theorycrafting.

This is better than we've got - so lets go with it 👍

**Review by m-samiec — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:33]: I looked at it and it wants a buffer - I would feel kind of bad casting the guid to a buffer with some length.
Though the current thing is also not ideal...

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:33]: It's totally fine to cast it, the char* is just so it can do pointer arithmetic on it. Guid internally even uses it for it's hash, but casts to a uint32 first. `CityHash64((char*)&Guid, sizeof(FGuid))`

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp]: Did it now

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp (hunk 2 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp]: fair enough

**Review by simonsarginson — APPROVED** [review]
